# Лабораторна робота №1. Лінійна регресія.

### Виконала студентка групи КМ-01мн Григоренко Анастасія

In [7]:
import pandas as pd
import numpy as np
from math import sqrt
from numpy import linalg 
from scipy.stats import t as ts

In [8]:
pd.set_option('display.max_rows', None)
df=pd.read_csv("MN-03-Hryhorenko.csv", sep = ';', header = None)
df.rename(columns = {0:'x1',1:'x2',2:'x3',3:'x4',4:'x5',5:'x6',6:'x7',7:'y'}, inplace = True)

In [11]:
Y = df[['y']]

In [12]:
df_test = df.iloc[200:250]

In [13]:
Y_train = Y.iloc[0:200]
Y_test = Y.iloc[200:250]

In [14]:
X = df[['x1','x2','x3','x4','x5','x6','x7']]

In [15]:
values=[]
for i in range(250):
    values.append(1)
X.insert(0,'x0', values, allow_duplicates=False)

In [16]:
X_train = X.iloc[0:200]
X_test = X.iloc[200:250]

In [17]:
X1 = df[['x1']]
X1.insert(0,'x0', values, allow_duplicates=False)
X1_train = X1.iloc[0:200]
X1_test = X1.iloc[200:250]

## Метод найменших квадратів

In [19]:
Xt=X_train.transpose()
b = (np.linalg.inv(Xt.dot(X_train))).dot(np.dot(Xt,Y_train))
b

array([[ 3.84887222],
       [-2.84384744],
       [ 9.52540507],
       [ 3.61724474],
       [ 1.38792175],
       [-4.8348413 ],
       [-3.33686675],
       [12.58074713]])

### Визначення коефіцієнту детермінації та перевірка статистичної значимості отриманого рівняння регресії

In [20]:
y_mean=Y_train.mean()

In [21]:
n = 200
Qr = float(np.dot(np.dot(np.transpose(b), Xt), Y_train)) - n*(y_mean**2)
Q = float(np.dot(np.transpose(Y_train), Y_train)) - n*(y_mean**2)

In [22]:
#Коефіцієнт кореляції
R = sqrt(Qr/Q)
R

0.8665464291864959

In [23]:
#Коефіцієнт детермінації
R2 = R**2
R2

0.7509027139358668

In [24]:
#Скоригований коефіцієнт детермінації
R_adj = 1 - ((n-1)/(n-7-1))*(1-R2)
R_adj

0.741821042048112

In [25]:
#Перевірка статистичної значимості
F = (R2*(n-7-1))/((1-R2)*7)
F

82.68331241391091

Порівнюємо отримане значення з табличним значенням F-критерію Фішера при рівні значущості 0.05. Враховуючи степені свободи k1=7 та k2=192, отримуємо Fтеор=2.135, що є меньшим від розрахованого. Отже, побудоване рівняння регресії є значимим, тобто залежна змінна Y достатньо добре описується включеними у регресійну модель змінними x1,x2,x3,x4,x5,x6,x7.

### Обчислення похибки збудованої моделі 

In [26]:
#Похибка
Y_predict = X_test.dot(b)

In [27]:
E2=(Y_test.T.to_numpy()-Y_predict.T.to_numpy())**2
Sigma = sqrt(E2.sum()/42)
Sigma

1.4036405341522913

Похибка отриманої регресійної моделі була обрахована в якості значення σ-середньоквадратичної помилки оцінки, тобто середньоквдратичного відхилення обрахованих даних від фактичних. σ вимірюється в тих самих одиницях виміру, що і Y. Тобто в середньому Y-обраховане відхиляється від Y-фактичне на 1.4 одиниці.


### Поелементні похибки

In [28]:
df_test['MNK'] = np.dot(X1_test,b)
df_test['Delta MNK'] = np.abs(df_test['y'] - df_test['MNK'] )
df_test

ValueError: shapes (50,2) and (8,1) not aligned: 2 (dim 1) != 8 (dim 0)

### Статистична значимість коефіцієнтів

In [29]:
koefs = b.T[0].copy()
Z = np.dot(X_train.transpose(), X_train)
z = linalg.inv(Z).diagonal()

In [30]:
Y_train['MNK'] = np.dot(X_train, koefs) 
s = ((Y_train['MNK'] - Y_train['y'])**2).sum()/(n-7-1)
sb = (s*z)**(1/2)
t = koefs/sb

C:\Users\Nastya\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [33]:
t_tabl = 0.3984
for ti, bi in zip(t, b):
    if np.abs(ti) < t_tabl:
        print(f'коеффициент {bi} статистически не значим')
    else:
        print(f'коеффициент {bi} статистически значим')

коеффициент [3.84887222] статистически значим
коеффициент [-2.84384744] статистически значим
коеффициент [9.52540507] статистически значим
коеффициент [3.61724474] статистически значим
коеффициент [1.38792175] статистически значим
коеффициент [-4.8348413] статистически значим
коеффициент [-3.33686675] статистически значим
коеффициент [12.58074713] статистически значим


## Градієнтний спуск

In [34]:
w_current= np.array([0,0,0,0,0,0,0,0])
# w_next = np.array([0,0,0,0,0,0,0,0])

i = 1
while True:
    n = 3/i
    Grad = (2/200)*(np.dot(Xt, np.transpose(np.dot(X_train, w_current)-Y_train.T.to_numpy())))
    w_next = (w_current - n*Grad.T)[0]
    if np.linalg.norm(w_current-w_next) < 0.0001:
        break
    w_current = w_next
    i += 1
#     print(w_current)

In [35]:
w = w_current
w

array([ 4.43651331,  2.08206035,  1.72897503,  0.60390583,  1.93007125,
       -0.10017445,  0.36378047,  1.09937891])

### Обчислення похибки збудованої моделі 

In [36]:
Y_predict_grad = X_test.dot(w)

In [37]:
E2_grad=(Y_test.T.to_numpy()-Y_predict_grad.T.to_numpy())**2
Sigma_grad = sqrt(E2_grad.sum()/42)
Sigma_grad

1.4128970253741735

Похибка отриманої регресійної моделі була обрахована в якості значення σ-середньоквадратичної помилки оцінки, тобто середньоквдратичного відхилення обрахованих даних від фактичних. σ вимірюється в тих самих одиницях виміру, що і Y. Тобто в середньому Y-обраховане відхиляється від Y-фактичне на 1.41 одиницю.

### Поелементні похибки 

In [38]:
df_test['Grad'] = np.dot(X_test,w)
df_test['Delta Grad'] = np.abs(df_test['y'] - df_test['Grad'] )
df_test

C:\Users\Nastya\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\Nastya\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


x1     x2     x3      x4      x5      x6      x7   y       Grad  \
200  0.670  0.510  0.180  1.6800  0.9260  0.2975  0.3935  13  10.512563   
201  0.675  0.550  0.190  1.5510  0.7105  0.3685  0.4120  13  10.416946   
202  0.680  0.520  0.165  1.4775  0.7240  0.2790  0.4060  11  10.178022   
203  0.680  0.530  0.180  1.5290  0.7635  0.3115  0.4025  11  10.307788   
204  0.700  0.525  0.175  1.7585  0.8745  0.3615  0.4700  10  10.861993   
205  0.700  0.550  0.200  1.5230  0.6930  0.3060  0.4405  13  10.431344   
206  0.725  0.530  0.190  1.7315  0.8300  0.3980  0.4050  11  10.825913   
207  0.725  0.550  0.200  1.5100  0.8735  0.4265  0.5085   9  10.558816   
208  0.735  0.570  0.175  1.8800  0.9095  0.3870  0.4880  11  11.272732   
209  0.740  0.575  0.220  2.0120  0.8915  0.5265  0.4710  12  11.607594   
210  0.750  0.555  0.215  2.2010  1.0615  0.5235  0.5285  11  12.000692   
211  0.190  0.140  0.030  0.0315  0.0125  0.0050  0.0105   3   5.165186   
212  0.210  0.150  0.045  0.0400  0.0135  0.0080  0.0105   4   5.250572   
213  0.250  0.175  0.060  0.0635  0.0275  0.0080  0.0200   4   5.440536   
214  0.290  0.215  0.065  0.0985  0.0425  0.0210  0.0310   5   5.678869   
215  0.335  0.250  0.080  0.1670  0.0675  0.0325  0.0575   6   6.005157   
216  0.340  0.245  0.085  0.2015  0.1005  0.0380  0.0530   6   6.070277   
217  0.345  0.255  0.095  0.1830  0.0750  0.0385  0.0600   6   6.078742   
218  0.355  0.255  0.080  0.1870  0.0780  0.0505  0.0580   7   6.100090   
219  0.360  0.260  0.080  0.1795  0.0740  0.0315  0.0600   5   6.100358   
220  0.370  0.275  0.090  0.2065  0.0960  0.0395  0.0580   7   6.203772   
221  0.375  0.290  0.140  0.3000  0.1400  0.0625  0.0825   8   6.481668   
222  0.375  0.275  0.095  0.2295  0.0950  0.0545  0.0660   7   6.275945   
223  0.385  0.300  0.125  0.3430  0.1705  0.0735  0.0810   7   6.593010   
224  0.385  0.285  0.085  0.2440  0.1215  0.0445  0.0680   8   6.331909   
225  0.395  0.320  0.100  0.3075  0.1490  0.0535  0.0900   8   6.569567   
226  0.400  0.305  0.100  0.3415  0.1760  0.0625  0.0865   7   6.616387   
227  0.405  0.305  0.100  0.2710  0.0965  0.0610  0.0910   7   6.503092   
228  0.405  0.310  0.110  0.9100  0.4160  0.2075  0.0995   8   7.781725   
229  0.405  0.305  0.100  0.2680  0.1145  0.0530  0.0850   7   6.485992   
230  0.405  0.300  0.090  0.2885  0.1380  0.0635  0.0765   6   6.502996   
231  0.410  0.315  0.100  0.3000  0.1240  0.0575  0.1000   8   6.592631   
232  0.410  0.325  0.110  0.3260  0.1325  0.0750  0.1010   8   6.672756   
233  0.415  0.335  0.100  0.3580  0.1690  0.0670  0.1050   7   6.754010   
234  0.420  0.325  0.115  0.3140  0.1295  0.0635  0.1000   8   6.668452   
235  0.420  0.315  0.110  0.4025  0.1855  0.0830  0.1015   8   6.822087   
236  0.430  0.340  0.110  0.3645  0.1590  0.0855  0.1050   7   6.820202   
237  0.445  0.360  0.110  0.4235  0.1820  0.0765  0.1400   9   7.032787   
238  0.450  0.325  0.115  0.4305  0.2235  0.0785  0.1155   8   6.968848   
239  0.450  0.335  0.095  0.3505  0.1615  0.0625  0.1185   7   6.823343   
240  0.455  0.340  0.115  0.4860  0.2610  0.0655  0.1315   8   7.121416   
241  0.460  0.350  0.100  0.4710  0.2520  0.0770  0.1230   8   7.106847   
242  0.460  0.345  0.105  0.4150  0.1870  0.0870  0.1100   8   6.988995   
243  0.475  0.355  0.115  0.5195  0.2790  0.0880  0.1325   7   7.261131   
244  0.480  0.375  0.120  0.5895  0.2535  0.1280  0.1720  11   7.504776   
245  0.485  0.380  0.125  0.5215  0.2215  0.1180  0.1600   8   7.381982   
246  0.485  0.365  0.140  0.4475  0.1895  0.0925  0.2305   8   7.293716   
247  0.490  0.365  0.125  0.5585  0.2520  0.1260  0.1615  10   7.439374   
248  0.505  0.385  0.125  0.5960  0.2450  0.0970  0.2100   9   7.621033   
249  0.505  0.380  0.135  0.5385  0.2645  0.0950  0.1650   9   7.455295   

     Delta Grad  
200    2.487437  
201    2.583054  
202    0.821978  
203    0.692212  
204    0.861993  
205    2.568656  
206    0.174087  
207    1.558816  
208    0.272732  
20